scalalib/src/mill/scalalib/CoverageModule.scala

```diff
@@ -0,0 +1,309 @@
+  * }
+  * }}}
+  *
+  * */
+trait CoverageModule extends ScalaModule with CoverageOptions {
+  import CoverageModule._
+  import CoverageUtil._
+
+  final override def ivyDeps =
+    super.ivyDeps() ++ Agg(ivyScoverageRuntime) ++ testIvyDeps()
+
+  def testIvyDeps = T { Agg.empty[Dep] }
+
+  override def compileIvyDeps = super.compileIvyDeps() ++ Agg(ivyScoverageRuntime)
+
+  override def scalacPluginIvyDeps =
+    super.scalacPluginIvyDeps() ++ Agg(ivyScoveragePlugin)
+
+  override def scalacOptions = T {
+    val pluginPath = {
+      val cp = scalacPluginClasspath()
+      cp.find { pathRef =>
+        val fileName = pathRef.path.toNIO.getFileName.toString
+        fileName.toString.contains(ScalacPluginArtifact) && fileName.toString
+          .contains(scoverageVersion)
+      }
+        .getOrElse {
+          throw new Exception(
+            s"Fatal: ${ScalacPluginArtifact} (${scoverageVersion}) not found on the classpath:\n\t" + cp
+              .map(_.path)
+              .mkString("\n\t"))
+        }
+    }.path
+    super.scalacOptions() ++ Seq(
+      Some(s"-Xplugin:${pluginPath}"),
+      Some(s"-P:scoverage:dataDir:${coverageDataDir().toNIO}"),
+      Option(excludedPackages).map(v => s"-P:scoverage:excludedPackages:$v"),
+      Option(excludedFiles).map(v => s"-P:scoverage:excludedFiles:$v"),
+      if (highlighting) Some("-Yrangepos") else None // rangepos is broken in some releases of scala so option to turn it off
+    ).flatten
+  }
+
+  private def coverageRootDir = T {
+    // FIXME: using artifactName is wrong
+    val path = millSourcePath / 'out / artifactName()
+      .split('-')
+      .last / scalaVersion() / 'coverage / 'dest
+    mkdir(path)
+    path
+  }
+
+  private def coverageDataDir = T {
+    val path = coverageRootDir() / 'data
+    mkdir(path)
+    path
+  }
+
+  def report = T {
+    val log = T.ctx.log
+    val dataDir = coverageDataDir().toNIO
+    val rootDir = coverageRootDir().toNIO
+    val encoding: Option[String] = sourceEncoding(scalacOptions())
+    val compileSourceDirectories =
+      sources().map(_.path.toNIO.toFile).filter(_.isDirectory)
+    log.info("reporting...")
+    loadCoverage(dataDir, log) match {
+      case Some(cov) =>
+        writeReports(rootDir,
+          cov,
+          this,
+          compileSourceDirectories,
+          encoding,
+          log)
+        checkCoverage(cov, log, this)
+      case None => log.error("No coverage data, skipping reports")
+    }
+  }
+}
+
+trait AggregateCoverageModule extends Module with CoverageOptions {
+  import CoverageUtil._
+
+  def modules: Seq[TestModule]
+
+  // FIXME: I want the sources!
+  def sources = T { modules.flatMap { module => module.sources() } }
+
+  def aggregate = T {
+    val log = T.ctx.log
+    log.info(s"Aggregating coverage from subprojects...")
+
+    val rootDir = (millSourcePath / 'dest).toNIO
+    val encoding: Option[String] = None // FIXME: sourceEncoding(scalacOptions())
```

Comment:
Ditto, how do I get the source encoding across modules (I can make sure they are all the same if necessary).

Reply:
See [transitiveIvyDeps](https://github.com/lihaoyi/mill/blob/master/scalalib/src/mill/scalalib/JavaModule.scala#L84)